standards/application/relay-reliability.md

```diff
@@ -0,0 +1,192 @@
+- **delivered**, the message is acknowledged by the receiver
+
+Application like Status already uses [MVDS](https://github.com/vacp2p/rfc-index/blob/main/vac/2/mvds.md) for e2e acknowledgement in direct messages and group chat. There is an ongoing [discussion](https://forum.vac.dev/t/end-to-end-reliability-for-scalable-distributed-logs/293) about a more general and bandwidth efficient solution for e2e reliablity.
+
+In other words, an application defines a payload over Waku and is interested in e2e delivery between application users. Waku provides a pub/sub broadcast transport, which is interested in reliably routing a message to all participants in the broadcast group.
+
+Before we have a complete design for e2e reliability, we need to compose existing protocols to increase the reliability of the transport protocol. This document proposes a few options for such composition. 
+
+## Motivation
+
+The [Store protocol](https://github.com/waku-org/specs/blob/master/standards/core/store.md) provides a way for nodes in the network to query the existence of messages or fetch specific messages based on the search criteria.
+
+**Search criteria with message hash**
+
+For the nodes that may have connection issues to **publish** messages via transport layer, this search criteria can be used to check whether a message is populated in the network or not. The message exists in store node can be marked from `outgoing` to `sent` by application. If the message is not found in the store node, the application can resend the message.
+
+**Search criteria with topics and time range**
+
+For the nodes that may have connection issues to **receive** messages via transport layer, this search criteria can be used to fetch missing messages from store nodes periodically after network resumes. 
+
+In summary, by leveraging the store node to provide such query services, the applications are able to mitigate the reliability issue of transport layer. This approach also introduces new challenges like centralization, privacy, and scalability. It should be viewed as a temporary solution and deprecated when e2e reliability solution is ready.
+
+## Implementation Suggestions 
+
+### Query with Message Hash
+
+For outgoing messages, the processing flow can be like this:
+- create a buffer for all "outgoing" messages
+- send message via relay or lightpush protocol
+- add message hash to the buffer
+- save the message to local database with status "outgoing"
+- check the buffer periodically
+- query the store node with message hash in the buffer, the messages should be posted more than a few seconds ago
+- if the message exists, update the status to "sent" and remove the message hash from the buffer
+- if the message does not exist, resend the message
+- if the message is still missing in the store node for a period of time, trigger the message failed to send workflow and remove the message hash from the buffer
+
+The implementation in Python may look like this:
```

Comment:
I would still move all the implementation suggestions to a separate section at the end of the spec. The heuristics can then belong to a single "Protocol" subheader and have subheadings that are more descriptive, e.g.

```
## Protocol

### Outgoing message reliability

### Incoming message reliability

### Unified approach

## Implementation suggestion

### Query with Message Hash

### Query with Topics and Time Range

```